difficulty/difficulty_test.go

```diff
@@ -254,36 +243,46 @@ func TestReciprocal(t *testing.T) {
 	}
 }
 
-// test decay
-func TestDecay(t *testing.T) {
+func TestIsAdjustBlockWhenStrtInterval(t *testing.T) {
+	height := uint64(difficulty.AdjustTimespanInBlocks * 200000)
+	actual := difficulty.IsAdjustBlock(height)
+	assert.Equal(t, true, actual, "fail to check starting of difficulty timespan")
+}
+
+func TestIsAdjustBlockWhenMiddleInterval(t *testing.T) {
+	height := uint64(difficulty.AdjustTimespanInBlocks*200000 + 1)
+	actual := difficulty.IsAdjustBlock(height)
+	assert.Equal(t, false, actual, "fail to check middle of difficulty timespan")
+}
 
-	initialReciprocal := 217.932
+func TestPrevTimespanBlockBeginAndEndWhenAtMiddle(t *testing.T) {
+	height := uint64(difficulty.AdjustTimespanInBlocks*3 + 10)
+	begin, end := difficulty.PrevTimespanBlockBeginAndEnd(height)
 
-	d := difficulty.New()
-	d.SetReciprocal(initialReciprocal)
+	assert.Equal(t, uint64(difficulty.AdjustTimespanInBlocks*3-1-difficulty.AdjustTimespanInBlocks), begin, "fail to get begin block")
+	assert.Equal(t, uint64(difficulty.AdjustTimespanInBlocks*3-1), end, "fail to get end block")
+}
 
-	percent := 100.0
+func TestPrevTimespanBlockBeginAndEndWhenAtStart(t *testing.T) {
+	height := uint64(difficulty.AdjustTimespanInBlocks * 3)
+	start, end := difficulty.PrevTimespanBlockBeginAndEnd(height)
 
-	// loop for three half lives to test at 50% 25% and 12.5%
-	for j := 0; j < 3; j += 1 {
+	assert.Equal(t, uint64(difficulty.AdjustTimespanInBlocks*3-1-difficulty.AdjustTimespanInBlocks), start, "fail to get begin block")
+	assert.Equal(t, uint64(difficulty.AdjustTimespanInBlocks*3-1), end, "fail to get end block")
+}
 
-		// this decay the value by one half life
-		for i := 0; i < difficulty.HalfLife; i += 1 {
-			d.Decay()
-		}
+func TestPrevTimespanBlockBeginAndEndWhenInFirstTimespan(t *testing.T) {
+	height := uint64(difficulty.AdjustTimespanInBlocks + 10)
+	begin, end := difficulty.PrevTimespanBlockBeginAndEnd(height)
 
-		percent /= 2.0
-		expected := initialReciprocal * percent / 100.0
-		actual := d.Reciprocal()
-
-		// the decay is fairly coarse so allow a resoable tolerance before issuing error
-		diff := actual - expected
-		diffPercent := 100.0 * diff / initialReciprocal
-		if math.Abs(diffPercent) > 0.5 {
-			t.Errorf("%5.2f%%: actual: %f  expected: %f", percent, actual, expected)
-			t.Errorf("%5.2f%%: diff: %g", percent, diff)
-			t.Errorf("%5.2f%%: relative: %6.2f%%", percent, 100*actual/initialReciprocal)
-			t.Errorf("%5.2f%%: error:    %6.2f%%", percent, diffPercent)
-		}
-	}
+	assert.Equal(t, uint64(2), begin, "fail to get begin block")
+	assert.Equal(t, uint64(difficulty.AdjustTimespanInBlocks-1), end, "fail to get end block")
+}
+
+func TestHashrate(t *testing.T) {
+	difficulty.Current.Set(4)
+	actual := difficulty.Hashrate()
+
+	expected := 8.533 // 1024 / 120
+	assert.Equal(t, expected, actual, "error calculate network hashrate")
```

Comment:
why is thin not a t.Fatalf showing values?

Reply:
the package will show expected and actual as follows:
```
--- FAIL: TestHashrate (0.00s)
    difficulty_test.go:288:
        	Error Trace:	difficulty_test.go:288
        	Error:      	Not equal:
        	            	expected: 8.433
        	            	actual  : 8.533
        	Test:       	TestHashrate
        	Messages:   	error calculate network hashrate
```